### ISSUE TITLE
[RFC] Plans for LLM QAT

### ISSUE THREAD

**Issue by andrewor14** [pytorch/ao#86]
Following the recent success of the [LLM-QAT](https://arxiv.org/pdf/2305.17888.pdf) paper, our high-level goal is to provide a PyTorch native workflow for LLM quantization-aware training (QAT) leveraging the quantization primitives and kernels provided by torchao, which is planned to become the de facto OSS library for AO techniques and kernels in PyTorch across different platforms (<https://github.com/pytorch-labs/ao/issues/47>). We also hope to eventually integrate with [TorchTune](https://github.com/pytorch/torchtune), a recently open-sourced library for fine-tuning and experimenting with LLMs, to provide an end-to-end flow that supports both finetuning and QAT.

## Workstream 1: Edge Devices<a id="workstream-1-edge-devices"></a>

Executorch provides a mechanism for [quantizing Llama2](https://github.com/pytorch/executorch/tree/main/examples/models/llama2) using post-training quantization (PTQ) techniques such as [GPTQ](https://arxiv.org/pdf/2210.17323.pdf), and lowering it to backends like XNNPACK. The main goal of this workstream is to provide a QAT drop-in replacement for GPTQ but with superior accuracy, starting with Llama2 7b using the following quantization/training configurations:

- **Linear weights:** 4-bit per channel grouped symmetric static quantization
- **Linear activations:** 8-bit per token symmetric dynamic quantization
- **Not “data-free”:** Use original dataset unlike in the LLM-QAT paper

We plan to adopt the same [eager mode quantization implementation](https://github.com/pytorch-labs/ao/blob/bed4cb461cba8ba4fee11b20d33528f272415dfd/torchao/quantization/quant_api.py#L630) used by the PTQ flow. In the future, if we decide to experiment with static quantization for activations for example, then we can explore using the [PT2 Export QAT flow](https://pytorch.org/tutorials/prototype/quantization_in_pytorch_2_0_export_tutorial.html).

## Workstream 2: Explore new quantization methods<a id="workstream-2-explore-new-quantization-methods"></a>

This workstream is largely backend agnostic; our goal is to motivate the backends (mobile or server CPU/GPU) to build the relevant kernels once we have demonstrated the initial success of a particular quantization configuration. There is a large design space we can experiment with summarized below. The suggested quantization and training techniques are primarily motivated by the [LLM-QAT](https://arxiv.org/abs/2305.17888) paper, but also by ongoing developments across the industry.

We can start with the following dimensions:

- **KV-cache quantization:** 4- or 8-bit KV-cache quantization can alleviate throughput bottlenecks with long sequences, and this has been shown (in the LLM-QAT paper) to hurt QAT a lot less than PTQ in terms of accuracy.

* **Custom dtypes:** The latest Hopper and upcoming Blackwell GPU generations [**no longer support int4 tensor cores**](https://www.nvidia.com/en-us/data-center/tensor-cores/), and so int4 kernels may not be as performant as other 4-bit dtypes. For example, both [**nf4**](https://arxiv.org/pdf/2305.14314.pdf) and [**MX4**](https://arxiv.org/pdf/2302.08007.pdf) promise higher fidelity than any a priori fixed quantization like int4. Experimenting with newer dtypes in QAT may lead to further accuracy gains.

- **Lower bit-widths:** 2- or 3-bit weight quantization can help further lower memory footprint and speed up inference. There have been PTQ attempts at such bit-widths (e.g. [**Quip#**](https://arxiv.org/pdf/2402.04396.pdf), [**AQLM**](https://arxiv.org/pdf/2401.06118.pdf)), but QAT has the potential to further mitigate the accuracy degradation.

## Workstream 3: Server GPU Inference<a id="workstream-3-server-gpu-inference"></a>

This is an extension of the recent [gpt-fast efforts](https://pytorch.org/blog/accelerating-generative-ai-2/) to quantize Llama but for QAT. An important goal here is to reuse the same quantization primitives as Workstream 1 to unify the two flows as much as possible. We can start with the following quantization configurations:

- **Int4 weight-only quantization.** This was the focus last half for Llama2, which targeted batch size 1 local chat agent use cases. This particular workload is memory bound, not compute bound, when run on GPUs, and so quantizing the activations here may not be particularly beneficial. For QAT, we can perform the same weight-only quantization for better accuracy.

* **Int4 weight quantization + int8 activation dynamic quantization,** similar to Workstream 1. One advantage here is we will have numerical baselines from the ExecuTorch workstream to compare against. However, as explained above, it may not make sense for Llama2 batch size 1 use cases for GPUs, so this configuration may be more suitable for larger batch sizes or other more compute bound models. The plan here is to be able to leverage ongoing efforts to provide mixed 4-bit / 8-bit GEMM in cutlass: <https://github.com/NVIDIA/cutlass/pull/1413>.

- **MX4 weight + activation quantization.** Please see the previous section under Custom dtypes for more details.
